spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveViews.scala

```diff
@@ -0,0 +1,156 @@
+import org.apache.spark.sql.internal.SQLConf
+
+case class ResolveViews(spark: SparkSession) extends Rule[LogicalPlan] with LookupCatalog {
+
+  import org.apache.spark.sql.connector.catalog.CatalogV2Implicits._
+
+  protected lazy val catalogManager: CatalogManager = spark.sessionState.catalogManager
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+    case u@UnresolvedRelation(nameParts, _, _) if catalogManager.v1SessionCatalog.isTempView(nameParts) =>
+      u
+
+    case u@UnresolvedRelation(
+    parts@NonSessionCatalogAndIdentifier(catalog, ident), _, _) if !isSQLOnFile(parts) =>
+      loadView(catalog, ident)
+        .map(createViewRelation(parts.quoted, _))
+        .getOrElse(u)
+
+    case p@SubqueryAlias(_, view: IcebergView) =>
+      p.copy(child = resolveViews(view))
+
+  }
+
+  def loadView(catalog: CatalogPlugin, ident: Identifier): Option[View] = catalog match {
+    case viewCatalog: ViewCatalog =>
+      try {
+        Option(viewCatalog.loadView(ident))
+      } catch {
+        case _: NoSuchViewException => None
+      }
+    case _ => None
+  }
+
+  // The current catalog and namespace may be different from when the view was created, we must
+  // resolve the view logical plan here, with the catalog and namespace stored in view metadata.
+  // This is done by keeping the catalog and namespace in `AnalysisContext`, and analyzer will
+  // look at `AnalysisContext.catalogAndNamespace` when resolving relations with single-part name.
+  // If `AnalysisContext.catalogAndNamespace` is non-empty, analyzer will expand single-part names
+  // with it, instead of current catalog and namespace.
+  def resolveViews(plan: LogicalPlan): LogicalPlan = plan match {
+    // The view's child should be a logical plan parsed from the `desc.viewText`, the variable
+    // `viewText` should be defined, or else we throw an error on the generation of the View
+    // operator.
+    case view@IcebergView(CatalogTableViewDescription(desc), isTempView, child) if !child.resolved =>
+      // Resolve all the UnresolvedRelations and Views in the child.
+      val newChild = AnalysisContext.withAnalysisContext(desc) {
+        val nestedViewDepth = AnalysisContext.get.nestedViewDepth
+        val maxNestedViewDepth = AnalysisContext.get.maxNestedViewDepth
+        if (nestedViewDepth > maxNestedViewDepth) {
+          throw QueryCompilationErrors.viewDepthExceedsMaxResolutionDepthError(
+            desc.identifier, maxNestedViewDepth, view)
+        }
+        SQLConf.withExistingConf(IcebergView.effectiveSQLConf(desc.viewSQLConfigs, isTempView)) {
+          execute(child)
+        }
+      }
+      // Fail the analysis eagerly because outside AnalysisContext, the unresolved operators
+      // inside a view maybe resolved incorrectly.
+      checkAnalysis(newChild)
+      view.copy(child = newChild)
+    case p@SubqueryAlias(_, view: IcebergView) =>
+      p.copy(child = resolveViews(view))
```

Comment:
Rules should not make recursive calls. If the rule needs to be applied more than once, it should happen by executing the rule multiple times.